### DISCUSSION TITLE
About global macro definitions

### DISCUSSION THREAD

**Discussion by tuopu** [openthread/openthread#7574] [Q&A] [answered]
Ask the expert for a question. 
I want to migrate to a new hardware platform, but the compilation failed. Do you have a more comprehensive migration description.
Is there a default definition independent of hardware that I can add hardware related drivers one by one to realize rapid migration？
I want to put the global macro definition of thread into a header file. Which file is more appropriate, such as "openthread_ftd = 1".
There is also the need to define those global macros when and whether there is a description file. thank you!!

**Comment by jwhui** [comment, marked as answer]
The build system typically defines this value depending on what it is trying to build.

Some examples:

- https://github.com/openthread/openthread/blob/237c91b9398bfe55426c45cbe5bcb5713c178ba3/Android.mk#L53
- https://github.com/openthread/openthread/blob/237c91b9398bfe55426c45cbe5bcb5713c178ba3/BUILD.gn#L46
- https://github.com/openthread/openthread/blob/237c91b9398bfe55426c45cbe5bcb5713c178ba3/src/core/ftd.cmake#L32

**Comment by tuopu** [comment]
Hello, expert, thank you for your reply. I'm learning to apply thread to FreeRTOS. When referring to the example, I found that the thread in OT RTOS can't be upgraded to the latest version, and it can't be compiled successfully in Linux. Please ask your latest submitted thread version to apply in FreeRTOS. Do you have any good suggestions?

**Comment by tuopu** [comment]
One more question，Under the world mechanism of thead, how can we know that the thread is about to arrive, so as to use semaphores and other system functions to inform the task to continue running?

**Comment by jwhui** [comment]
> One more question，Under the world mechanism of thead, how can we know that the thread is about to arrive, so as to use semaphores and other system functions to inform the task to continue running?

Platform drivers handle events (e.g. timer fires, radio packets received, etc.) and those drivers are responsible for scheduling OpenThread execution at the appropriate time depending on the platform's implementation.
